app/src/main/java/pl/pwr/zpi/bcycle/mobile/api/ApiClient.kt

```diff
@@ -0,0 +1,65 @@
+package pl.pwr.zpi.bcycle.mobile.api
+
+import com.google.gson.GsonBuilder
+import com.google.gson.TypeAdapter
+import com.google.gson.stream.JsonReader
+import com.google.gson.stream.JsonToken
+import com.google.gson.stream.JsonWriter
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import org.threeten.bp.ZonedDateTime
+import pl.pwr.zpi.bcycle.mobile.API_BASE_URL
+import pl.pwr.zpi.bcycle.mobile.utils.dateFromIso
+import pl.pwr.zpi.bcycle.mobile.utils.dateToIso
+import retrofit2.Retrofit
+import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
+import retrofit2.converter.gson.GsonConverterFactory
+
+
+object ApiClient {
+    var currentToken: String = "" // updated by MainActivity
+
+    private val retrofit: Retrofit by lazy {
+        Retrofit.Builder()
+            .baseUrl(API_BASE_URL)
+            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
+            .addConverterFactory(GsonConverterFactory.create(gson))
+            .client(okHttpClient)
+            .build()
+    }
+
+    val tripApi: TripApi by lazy {
+        retrofit.create(TripApi::class.java)
+    }
+
+    private val gson = GsonBuilder().registerTypeAdapter(ZonedDateTime::class.java, object :
```

Comment:
A really don't want get into details as far as it works,  but at least declare that object as in separate file. I also saw something similar in RecordActivity.

Reply:
It’s an anonymous object required by Gson only, I don’t see the point in separating it out. The objects in RecordTripActivity cannot be moved, since they need to be able to use attributes of that activity (ServiceConnection comes from Android docs examples.)

Reply:
Despite the fact that's bad code, it just looks like terrible, if you don't see it, that's your problem. Do what you want, not gonna get excited over university project - it's a waste of time.